### PR TITLE
Coverage checking for ContractModels

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
This PR introduces coverage checking for QuickCheck ContractModels. We currently do two types of coverage:
* What endpoints have been called and how often and
* What source locations in the validators have been covered *on chain*

Pre-submit checklist:
- Branch
    - [x] Tests are provided (if possible)
    - [X] Commit sequence broadly makes sense
    - [X] Key commits have useful messages
    - [x] Relevant tickets are mentioned in commit messages
    - [x] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [x] Self-reviewed the diff
    - [x] Useful pull request description
    - [x] Reviewer requested

### CODE REVIEW

**Comment by MaximilianAlgehed** [comment]
I have *no clue* what is going on, but on my local machine I get the following completely befuddling build error when building this against `plutus-tx`:
```
src/Plutus/Contract/Test/Coverage.hs:19:1: error:
    Could not load module ‘PlutusTx.Coverage’
    It is a member of the hidden package ‘plutus-tx-0.1.0.0’.
    Perhaps you need to add ‘plutus-tx’ to the build-depends in your .cabal file.
    Use -v (or `:set -v` in ghci) to see a list of the files searched for.
   |
19 | import           PlutusTx.Coverage
   |
```

**Comment by sjoerdvisscher** [comment]
@MaximilianAlgehed Did the latest version fix your issues?

**Comment by MaximilianAlgehed** [comment]
@sjoerdvisscher it was a really odd issue and its been fixed now. We had the dependency pinned to the wrong thing in the `cabal.project` file. Which is my bad. However, the error message we got didn't help - that was from there being another version of the plutus repo *somewhere* in the nix universe that `cabal` decided it would be a good idea to try to pull in. In other words, it's no issue :)

**Comment by MaximilianAlgehed** [comment]
@sjoerdvisscher Any progress on reviewing this?

File: plutus-use-cases/src/Plutus/Contracts/GameStateMachine.hs [hunk 1]
```diff
@@ -13,7 +13,8 @@
 {-# LANGUAGE TypeOperators         #-}
 {-# LANGUAGE ViewPatterns          #-}
 {-# OPTIONS_GHC -fno-ignore-interface-pragmas #-}
-{-# OPTIONS_GHC -fplugin-opt PlutusTx.Plugin:debug-context #-}
+{-# OPTIONS_GHC -g -fplugin-opt PlutusTx.Plugin:coverage-all #-}
+-- You need to use all of these to get coverage
```
Comment (sjoerdvisscher) [inline plutus-use-cases/src/Plutus/Contracts/GameStateMachine.hs:17]:
```suggestion
-- You need to use all of these to get coverage

```

I think you need an extra empty line here to not break the haddock comment below it.

**Review by sjoerdvisscher — APPROVED** [review]
Awesome stuff!

**Comment by MaximilianAlgehed** [comment]
This is failing in code that this PR dosn't touch. I think this is an issue on master.

**Comment by sjoerdvisscher** [comment]
Indeed. I just fixed master.

**Comment by MaximilianAlgehed** [comment]
I don't know what's going on with hydra but I guess it will work itself out eventually.

**Comment by sjoerdvisscher** [comment]
@MaximilianAlgehed Somehow you missed the commit that fixes the build!

**Comment by MaximilianAlgehed** [comment]
@sjoerdvisscher I don't think there is anything I can do to make hydra restart and as far as I can tell this is fully rebased on `main`. What do?

**Pull request merged**